Allow IPv6 with NetworkInterfaceIPConfiguration

imrejonk opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

The ip_configurations parameter of the azure_rm_networkinterface module did not allow specifying the IP version, causing the Azure REST API to default to IPv4. Trying to add a public IPv6 address to the network interface would cause this error:

```
Error creating or updating network interface mynic - Azure Error: IPVersionForPublicIpAndNicIpConfigurationMustMatch
    Message: IP version for publicIPAddress '/subscriptions/[..]/resourceGroups/ansible/providers/Microsoft.Network/publicIPAddresses/mypublicipv6address', version 'IPv6' and network
    interface ipConfiguration '/subscriptions/[..]/resourceGroups/ansible/providers/Microsoft.Network/networkInterfaces/mynic/ipConfigurations/myipv6config', version 'IPv4' must match.
```

The Azure Python SDK however does [allow](https://docs.microsoft.com/en-us/python/api/azure-mgmt-network/azure.mgmt.network.v2020_06_01.models.networkinterfaceipconfiguration?view=azure-python) for a parameter called `private_ip_address_version`. This commit implements that parameter for the azure_rm_networkinterface module.

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Feature Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
azure_rm_networkinterface

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

You can now do this:

```
- name: Create network interfaces
  azure_rm_networkinterface:
    name: 'mynic'
    resource_group: 'myrg'
    location: 'westeurope'
    virtual_network: 'myvirtnet'
    subnet_name: 'mysubnet'
    ip_configurations:
      - name: 'myipv4config'
        primary: True
        public_ip_address_name: 'mypublicipv4address'
        private_ip_address_version: 'IPv4'
      - name: 'myipv6config'
        primary: False
        public_ip_address_name: 'mypublicipv6address'
        private_ip_address_version: 'IPv6'
```
And I can remove an ugly workaround involving an Ansible `command` task that just called azure-cli to add IPv6 configurations to network interfaces :blush:

imrejonk committed:
Allow IPv6 with NetworkInterfaceIPConfiguration

The ip_configurations parameter of the azure_rm_networkinterface module
did not allow specifying the IP version, causing the Azure REST API to
default to IPv4. Trying to add a public IPv6 address to the network
interface would cause this error:

```
Error creating or updating network interface mynic - Azure Error: IPVersionForPublicIpAndNicIpConfigurationMustMatch
    Message: IP version for publicIPAddress '/subscriptions/[..]/resourceGroups/ansible/providers/Microsoft.Network/publicIPAddresses/mypublicipv6address', version 'IPv6' and network
    interface ipConfiguration '/subscriptions/[..]/resourceGroups/ansible/providers/Microsoft.Network/networkInterfaces/mynic/ipConfigurations/myipv6config', version 'IPv4' must match.
```

The Azure Python SDK however does allow for a parameter called
`private_ip_address_version` [0]. This commit implements that parameter
for the azure_rm_networkinterface module.

[1] https://docs.microsoft.com/en-us/python/api/azure-mgmt-network/azure.mgmt.network.v2020_06_01.models.networkinterfaceipconfiguration?view=azure-python

Fred-sun commented:
@imrejonk Thanks for your contribution. Can you add test cases for the newly added feature? Thank you very much!

imrejonk commented:
No problem, I'll add some test cases next week. I just got service principal authentication working for ansible-test and am now doing some trial runs with the existing integration tests.

imrejonk commented:
@Fred-sun I've modified the `azure_rm_networkinterface` integration tests to account for this change and rebased my branch. I also fixed an idempotency issue related to the `primary` parameter, which was being set from null to False on the second run when the parameter was omitted.

Some things to note:
- I've set the public IP address SKU to 'standard', and the allocation method to 'static'. This is due to [limitations](https://docs.microsoft.com/en-us/azure/virtual-network/public-ip-address-prefix#limitations) in Azure's IPv6 support.
- The total number of public IP addresses being created now comes to four: two for the load balancer (IPv4 and IPv6), and two for the network interface.
- Mixing 'basic' with 'standard' SKUs in a load balancer or NIC configuration is not allowed, so I had to change how IPv4 public addresses were created in the tests.
- I noticed that network security groups and the virtual network were not being removed after running the tests, so I added tasks for cleaning them up.

Let me know if this is sufficient, and if not, what types of additional tests you would like to see.

Fred-sun commented on tests/integration/targets/azure_rm_networkinterface/tasks/main.yml:
```diff
@@ -10,31 +10,47 @@
       resource_group: "{{ resource_group_secondary }}"
       name: "tn{{ rpfx }}"
-      address_prefixes: "10.10.0.0/16"
+      address_prefixes: ["10.10.0.0/16","fdae:f296:2787::/48"]
```
```suggestion
      address_prefixes: ["10.10.0.0/16", "fdae:f296:2787::/48"]
```

Fred-sun commented on tests/integration/targets/azure_rm_networkinterface/tasks/main.yml:
```diff
@@ -10,31 +10,47 @@
       resource_group: "{{ resource_group_secondary }}"
       name: "tn{{ rpfx }}"
-      address_prefix: "10.10.0.0/24"
+      address_prefixes_cidr: ["10.10.0.0/24","fdae:f296:2787::/64"]
```
too few spaces after comma
```suggestion
      address_prefixes_cidr: ["10.10.0.0/24", "fdae:f296:2787::/64"]
```

imrejonk commented:
@Fred-sun here you go, style issues fixed.

Fred-sun commented:
> @Fred-sun here you go, style issues fixed.

@imrejonk Merger will be promoted in the near future. Thank you very much!

xuzhang3 commented:
LGTM
